R/functions.R

```diff
@@ -395,224 +382,260 @@ vwparams <- function(vwmodel, name) {
-  if(length(params$reductions) != 0) {
-      valid_reductions <- list()
-      params$reductions <- sapply(names(params$reductions), function(reduction_name) {
-          reduction_check_type <- get(paste0(reduction_name, "_check"))
-          valid_reduction <- check_param_values(
-              input = params$reductions[[reduction_name]],
-              check = reduction_check_type
-          )
-          valid_reduction <- setNames(list(valid_reduction), reduction_name)
-          valid_reductions <<- c(valid_reductions, valid_reduction)
-      })
-      params$reductions <- valid_reductions
+    # parse categorical variables
+    parsingVar <- function(x, keepSpace) {
+        # remove leading and trailing spaces, then remove special characters
+        # then remove isolated underscores.
+        if (!keepSpace)
+            spch <- specCharSpace
+        else
+            spch <- specChar
+        gsub(spch, '_', x)
+    }
     
-  }
-  if(length(params$general_params) == 0) {
-      params$general_params <- general_check
-  } else {
-      params$general_params <- check_param_values(
-          input = params$general_params,
-          # input = c(list(cache=params$cache), params$general_params),
-          check = general_check
-      )
-  }
-  if(length(params$optimization_params) == 0) {
-      algorithm_parameters <- get(paste0(params$algorithm, "_check"))
-      params$optimization_params <- c(algorithm_parameters, optimization_check)
-  } else {
-      algorithm_check_type <- get(paste0(params$algorithm, "_check"))
-      params$optimization_params <- check_param_values(
-          input = params$optimization_params,
-          check = c(algorithm_check_type, optimization_check)
-      )
-  }
-  
-  # Cache should be created, if passes > 1
-  if(params$general_params$passes > 1) {
-      params$general_params$cache <- TRUE
-  }
-  # Return validated parameters
-  return(list(algorithm = params$algorithm,
-              general_params = params$general_params,
-              optimization_params = params$optimization_params,
-              reductions = params$reductions))
-}
-
-.create_parameters_string <- function(params) {
-    params_to_strings <- function(i) {
-        if(is.na(flat_params[[i]]) | flat_params[[i]] == "") {
-            return("")
-        };
-        if(is.logical(flat_params[[i]][[1]]) & flat_params[[i]][[1]] == TRUE) {
-            return(paste0("--",i))
-        }; 
-        if(is.logical(flat_params[[i]][[1]]) & flat_params[[i]][[1]] == FALSE) {
-            return("")
-        } else {
-            return(paste0("--",i," ",flat_params[[i]]))
+    # Construct labels for multilabel examples: 1:0.3 2:0.3 3:0.3
+    constructLabels <- function(targets, num_row) {
+        # Initialize output vector
+        tmp <- rep("", num_row)
+        # Iterate cost vectors names
+        for(i in 1:length(targets)) {
+            elem_targets <- eval(get(targets[i], envir=parent.frame(n=1)))
```

Comment:
This is a _very_ dangerous proposition. Why are we reaching into the parent environment for a symbol and then evaluating it?

Reply:
I found this idea in previous version of converter in `rvw` package. I thought this was a good idea and decided to continue using it in my version.

Ok, I see now that I should rework this logic.
But why is this idea very dangerous?

Reply:
Sorry for missing the response on this. The reason it's dangerous is it assumes environment position for the elements is 1 above, the object has been created, and that it should be evaluated once retrieved (`eval()`).